storage/database/sharded_database.go

```diff
@@ -155,75 +161,251 @@ func (db *shardedDB) Close() {
 	}
 }
 
+// Not enough size of channel slows down the iterator
+const shardedDBCombineChanSize = 1024 // Size of resultCh
+const shardedDBSubChannelSize = 128   // Size of each sub-channel of resultChs
+
+// shardedDBIterator iterates all items of each shardDB.
+// This is useful when you want to get items in serial in binary-alphabetigcal order.
 type shardedDBIterator struct {
-	// TODO-Klaytn implement this later.
-	iterators []Iterator
-	key       []byte
-	value     []byte
+	shardedDBChanIterator
 
-	//numBatches uint
-	//
-	//taskCh   chan pdbBatchTask
-	//resultCh chan pdbBatchResult
+	resultCh chan common.Entry
+	key      []byte // current key
+	value    []byte // current value
 }
 
 // NewIterator creates a binary-alphabetical iterator over a subset
 // of database content with a particular key prefix, starting at a particular
 // initial key (or after, if it does not exist).
-func (pdb *shardedDB) NewIterator(prefix []byte, start []byte) Iterator {
-	// TODO-Klaytn implement this later.
-	return nil
+func (db *shardedDB) NewIterator(prefix []byte, start []byte) Iterator {
+	it := &shardedDBIterator{
+		shardedDBChanIterator: db.NewChanIterator(context.TODO(), prefix, start, nil),
+		resultCh:              make(chan common.Entry, shardedDBCombineChanSize),
+	}
+
+	go it.runCombineWorker()
+
+	return it
+}
+
+// runCombineWorker fetches any key/value from resultChs and put the data in resultCh
+// in binary-alphabetical order.
+func (it *shardedDBIterator) runCombineWorker() {
+
+	// creates min-priority queue smallest values from each iterators
+	entries := &entryHeap{}
+	heap.Init(entries)
+	for i, ch := range it.resultChs {
+		if e, ok := <-ch; ok {
+			heap.Push(entries, entryWithShardNum{e, i})
+		}
+	}
+
+chanIter:
+	for len(*entries) != 0 {
+		// check if done
+		select {
+		case <-it.ctx.Done():
+			logger.Trace("[shardedDBIterator] combine worker ends due to ctx")
+			break chanIter
+		default:
+		}
+
+		// look for smallest key
+		minEntry := heap.Pop(entries).(entryWithShardNum)
+
+		// fill resultCh with smallest key
+		it.resultCh <- minEntry.Entry
+
+		// fill used entry with new entry
+		// skip this if channel is closed
+		if e, ok := <-it.resultChs[minEntry.shardNum]; ok {
+			heap.Push(entries, entryWithShardNum{e, minEntry.shardNum})
+		}
+	}
+	logger.Trace("[shardedDBIterator] combine worker finished")
+	close(it.resultCh)
+}
+
+type entryWithShardNum struct {
+	common.Entry
+	shardNum int
+}
+
+type entryHeap []entryWithShardNum
```

Comment:
Is this `Heap` not `Stack`? Looks `Push` appends an item at the end of slice, and the `Pop` pops the last element.